subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java

```diff
@@ -811,73 +813,41 @@ public boolean apply(final SubscriptionBaseEvent input) {
-
-        if (!isCancelled) {
+    private void removeEverythingPastCancelEvent(final List<SubscriptionBaseEvent> inputEvents) {
+        final SubscriptionBaseEvent cancellationEvent = Iterables.tryFind(inputEvents,
+                                                                          new Predicate<SubscriptionBaseEvent>() {
+                                                                              @Override
+                                                                              public boolean apply(final SubscriptionBaseEvent input) {
+                                                                                  return input.getType() == EventType.API_USER && ((ApiEvent) input).getApiEventType() == ApiEventType.CANCEL;
+                                                                              }
+                                                                          }).orNull();
+        if (cancellationEvent == null) {
             return;
         }
 
-
-        boolean foundFirstOnDiskCancel = false;
-        final Iterator<SubscriptionBaseEvent> it =  inputEvents.iterator();
-        while(it.hasNext()) {
+        final Iterator<SubscriptionBaseEvent> it = inputEvents.iterator();
+        while (it.hasNext()) {
             final SubscriptionBaseEvent input = it.next();
             if (!input.isActive()) {
                 continue;
             }
 
-            if (input.getType() == EventType.API_USER) {
-                final ApiEvent userEV = (ApiEvent) input;
-                if (userEV.getApiEventType() == ApiEventType.CANCEL) {
-                    if (userEV.isFromDisk()) {
-                        if (!foundFirstOnDiskCancel) {
-                            foundFirstOnDiskCancel = true;
-                        } else {
-                            it.remove();
-                        }
-                    } else {
-                        it.remove();
-                    }
-                }
+            if (input.getId().compareTo(cancellationEvent.getId()) == 0) {
+                // Keep the cancellation event
+            } else if (input.getType() == EventType.API_USER && (((ApiEvent) input).getApiEventType() == ApiEventType.TRANSFER || ((ApiEvent) input).getApiEventType() == ApiEventType.CREATE)) {
+                // Keep the initial event (SOT use-case)
+            } else if (input.getType() == EventType.API_USER && (((ApiEvent) input).getApiEventType() == ApiEventType.CANCEL) && !((ApiEvent) input).isFromDisk()) {
+                // Also we remove !onDisk cancel events when there is an onDisk cancel event (can happen during the path where we process the base plan cancel notification, and are
+                // in the process of adding the new cancel events for the AO)
+                it.remove();
+            } else if (input.getEffectiveDate().compareTo(cancellationEvent.getEffectiveDate()) >= 0) {
+                // Event to ignore past cancellation date
+                it.remove();
```

Comment:
Trying to step back from what we had and thinking about what i think we should have:

Assumptions: events have been ordered using the `Comparator` above -- you fixed.

1. First  event should be `CREATE ` or `TRANSFER` => Delete every strictly before  this.
2. One we hit the first `CANCEL`, delete everything after or equals to this date. Note that because of the ordering, this should take into account use case where we have a `!isFromDisk` `CANCEL` event because it would come after in the stream and therefore would be deleted with this logic.

Also as we remove events from the stream --maybe it would be worth logging a `WARN` as this represent data issues i would think.

Reply:
> One we hit the first `CANCEL`, delete everything after or equals to this date. Note that because of the ordering, this should take into account use case where we have a `!isFromDisk` `CANCEL` event because it would come after in the stream and therefore would be deleted with this logic.

Right, I think this branch could be deleted now that the Dao bug has been fixed. I'll check.

> Also as we remove events from the stream --maybe it would be worth logging a `WARN` as this represent data issues i would think.

I don't think so, unless we change other parts of the code. For add-ons, for instance, we almost always have two `CANCEL` events, one from the disk and one added in memory.

Reply:
> Right, I think this branch could be deleted now that the Dao bug has been fixed. I'll check.

Confirmed in 221405dfcd02dbfc1e3858bc50fbcfe6f48d0a03.